### PR TITLE
HHH-10182 Backport HHH-7898 for JTA environment only

### PR DESCRIPTION
**Pull request by rvansa**
* this fix works only for JTA transactions! (JDBC transactions behavior was not altered)
* removed cluster loader from default configuration

### CODE REVIEW

File: hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java [hunk 1]
```diff
@@ -1,79 +1,209 @@
...
  * @author Galder Zamarreño
  * @since 3.5
  */
 public class QueryResultsRegionImpl extends BaseTransactionalDataRegion implements QueryResultsRegion {
+   private static final Log log = LogFactory.getLog( QueryResultsRegionImpl.class );
 
    private final AdvancedCache evictCache;
    private final AdvancedCache putCache;
    private final AdvancedCache getCache;
+   private final ConcurrentMap<Transaction, Map<Object, PostTransactionQueryUpdate> > transactionContext
```
Comment (galderz) [inline hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java]: Could you explain the reasons for the different in the value type for the Map here compared to master? In master the value is the query value passed in.

File: hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java [hunk 2]
```diff
@@ -1,79 +1,209 @@
...
+         // and the result would not be consistent.
+         TransactionManager tm = getTransactionManager();
+         Transaction transaction = null;
+         try {
+            transaction = tm != null && tm.getStatus() == Status.STATUS_ACTIVE ? tm.getTransaction() : null;
+            if (transaction != null) {
+               Map<Object, PostTransactionQueryUpdate> map = transactionContext.get(transaction);
+               if (map == null) {
+                  // The map implementation needs to be synchronized as we can access that from evict, too.
+                  transactionContext.put(transaction, map = new ConcurrentHashMap());
```
Comment (galderz) [inline hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java]: Why the need for ConcurrentHashMap vs Map in master? Is it due to the need to set value to null? I guess this goes back to the original question at the top on the difference of `transactionContext` map's map value type....

File: hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java [hunk 3]
```diff
@@ -89,8 +219,76 @@ public void put(Object key, Object value) throws CacheException {
...
+               TransactionManager tm = getTransactionManager();
+               Transaction suspended = null;
+               try {
+                  suspended = tm.suspend();
+                  if (putCacheRequiresTransaction) {
+                     tm.begin();
+                     try {
+                        putCache.put(key, value);
+                     } finally {
+                        tm.commit();
```
Comment (galderz) [inline hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java]: You also need to rollback this if necessary..., don't you? e.g.

     catch (Exception e) {
               tm.setRollbackOnly();
               throw e;
            } finally {
               if(tm.getStatus() == 0) {
                  tm.commit();
               } else {
                  tm.rollback();
               }
            }

File: hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java (hunk 1 above)
Comment (rvansa) [inline hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java]: When multiple puts are invoked in the same transaction, we need to apply only the last of them, therefore we update the synchronization. So the actual behaviour in master can be incorrect - there the value is overwritten in the cache (and further queries in the same transaction) return the correct value, but in the end we update the cache several times. Within Hibernate's transactional mechanism, that ends up fine - the synchronizations are invoked in the same order as these are registered, but JTA transactions do not guarantee this and in fact JBossTM invokes the afterCommit in reverse order.

File: hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java (hunk 3 above)
Comment (galderz) [inline hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java]: You already have `org.hibernate.cache.infinispan.util.Caches` to deal with similar patterns, don't you?

File: hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java (hunk 2 above)
Comment (rvansa) [inline hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java]: Again, flaw in master discovered during the review process. `evict()` should update the values in `transactionContext`, too, and therefore there can be concurrent access to it. As the comment one line above says.

**Comment by rvansa** [comment]
I have added a commit with the transaction wrap using Caches.withinTx.

**Comment by rvansa** [comment]
Galder thinks that I should squash the commits, so I've squashed them all into single commit that should go into 4.2

**Pull request merged**